Core/src/main/java/org/tribuo/protos/ProtoUtil.java

```diff
@@ -0,0 +1,474 @@
+        String tpName = ProtoSerializable.class.getTypeParameters()[0].getName();
+        throw new IllegalArgumentException("unable to resolve type parameter '" + tpName + "' in ProtoSerializable<" + tpName + "> for class " + protoSerializable.getClass().getName());
+    }
+
+    private static List<Object> convertValues(Map<?, ?> obj) {
+        List<Object> values = new ArrayList<>();
+        for (Object value : obj.values()) {
+            values.add(convert(value));
+        }
+        return values;
+    }
+
+    private static Object convert(Object obj) {
+        if (obj instanceof ProtoSerializable) {
+            return ((ProtoSerializable<?>) obj).serialize();
+        } else if (obj instanceof MutableLong) {
+            return ((MutableLong) obj).longValue();
+        } else if (obj instanceof MutableDouble) {
+            return ((MutableDouble) obj).doubleValue();
+        } else if (obj.getClass().isEnum()) {
+            return ((Enum<?>) obj).name();
+        } else if (obj instanceof int[]) {
+            int[] t = (int[]) obj;
+            return Arrays.stream(t).boxed().collect(Collectors.toList());
+        } else if (obj instanceof long[]) {
+            long[] t = (long[]) obj;
+            return Arrays.stream(t).boxed().collect(Collectors.toList());
+        } else if (obj instanceof float[]) {
+            float[] t = (float[]) obj;
+            List<Float> floats = new ArrayList<>();
+            for (float f : t) {
+                floats.add(f);
+            }
+            return floats;
+        } else if (obj instanceof double[]) {
+            double[] t = (double[]) obj;
+            return Arrays.stream(t).boxed().collect(Collectors.toList());
+        } else if (obj instanceof String[]) {
+            return Arrays.asList((String[]) obj);
+        } else {
+            return obj;
+        }
+    }
+
+    private static List<Field> getFields(Class<?> clazz) {
+        Set<String> fieldNameSet = new HashSet<>();
+        List<Field> fields = new ArrayList<>();
+        getFields(clazz, fieldNameSet, fields);
+        return fields;
+    }
+
+    private static void getFields(Class<?> clazz, Set<String> fieldNameSet, List<Field> fields) {
+        for (Field field : clazz.getDeclaredFields()) {
+            ProtoSerializableField psf = field.getAnnotation(ProtoSerializableField.class);
+            ProtoSerializableKeysValuesField pskvf = field.getAnnotation(ProtoSerializableKeysValuesField.class);
+            ProtoSerializableMapValuesField psmvf = field.getAnnotation(ProtoSerializableMapValuesField.class);
+            if ((psf != null) && (pskvf == null) && (psmvf == null)) {
+                String protoFieldName;
+                if (psf.equals(ProtoSerializableField.DEFAULT_FIELD_NAME)) {
+                    protoFieldName = field.getName();
+                } else {
+                    protoFieldName = psf.name();
+                }
+                if (!fieldNameSet.contains(protoFieldName)) {
```

Comment:
Isn't having a duplicated name here a big problem that shouldn't happen?

Reply:
Possibly, though Java allows field shadowing in subclasses. That said it should probably go bang if there is shadowing because it's likely to make the serialized format impossible to reason about.

Reply:
We should probably warn the user if we hit this, then.

Reply:
Done.